Add new datatype GRIB

annefou opened the pull request:
Added GRIB (GRIdded Binary or General Regularly-distributed Information in Binary form) to support meteorological data in Galaxy (used by Numerical Weather Prediction model outputs):
    - Can be identified with magic 'GRIB' and edition (1 or 2)
    - It is a binary data format so info on EDAM is binary but there is work in progress to add GRIB as a new data format in EDAM
    - Added test.grib to check 'sniff' function. I added a test but I did not manage to run the automated Galaxy tests but I check the 'sniff' function manually on my laptop with python 3.8.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

Commit:
add GRIB data format (meteorology)

Commit:
check GRIB edition too

Commit:
Added GRIB (GRIdded Binary or General Regularly-distributed Information in Binary form)
- Can be identified with magic 'GRIB' and check edition (1 or 2)
- Binary format so info on EDAM is binary but there is work in progress to add GRIB as a new data format in EDAM

Co-authored by: melibleq <melibleq@users.noreply.github.com>

bgruening commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -1224,6 +1224,56 @@ def display_peek(self, dataset):
+                tmp = f.read(4)
+                edition = _uint8struct.unpack_from(tmp, 3)[0]
+                if edition == 1 or edition == 2:
+                    return True
```
Is that edition important for you? If so you could save it as metadata.

annefou replied:
I added it because I was a bit worried to only check 'GRIB' but it is a good idea to save it as metadata. I am not sure how to do it properly. 
1. Add MetadataElement: 
` MetadataElement(name="grib_edition", default=1, desc="GRIB edition", readonly=True, visible=True, optional=True, no_value=0)`
2. Add `set_meta`:
```
def set_meta(self, dataset, **kwd):
        """
        Set the GRIB edition.
        """
        dataset.metadata.grib_edition = self._get_grib_edition(dataset.file_name)

def _get_grib_edition(self, filename):
         _uint8struct = struct.Struct(b">B")
         edition = 0
          with open(filename) as f:
               header = f.read(4)
               tmp = f.read(4)
               edition = _uint8struct.unpack_from(tmp, 3)[0]
           return edition
```

Would it work? It is sufficient or is there anything else to add? Thanks!

bgruening replied:
that looks ok to me

mvdbeek commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -1224,6 +1224,56 @@ def display_peek(self, dataset):
+            header = f.read(4)
+            if header == self._magic:
+                tmp = f.read(4)
+                edition = _uint8struct.unpack_from(tmp, 3)[0]
```
```suggestion
    def sniff_prefix(self, sniff_prefix):
        # The first 4 bytes of any GRIB file are GRIB
        try:
            if sniff_prefix.startswith_bytes(self._magic):
                tmp = contents_header_bytes[:4]
                _uint8struct = struct.Struct(b">B")
                edition = _uint8struct.unpack_from(tmp, 3)[0]
```
sniff_prefix is a small optimization that reads in the first portion of a file once, and then provides this to all sniff_prefix methods. This way we don't have to open the files to be sniffed N times for N datatypes that define a sniffer function.

Commit:
Use sniff_prefix to optimize

sniff_prefix is a small optimization that reads in the first portion of a file once, and then provides this to all sniff_prefix methods. This way we don't have to open the files to be sniffed N times for N datatypes that define a sniffer function.

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

Commit:
add metadata for GRIB edition

Commit:
fix issue with indentation

Commit:
Merge branch 'dev' into grib

Commit:
use contents_header_bytes and FilePrefix for GRIB datatype

Commit:
add missing FilePrefix in import for binary

Commit:
use tmp variable instead of header
because it is not used

Commit:
attempt to fix tests with GRIB datatype
not sure this is correct, mostly trying...

mvdbeek committed:
Fix grib sniffer

The initial code had two read(4) calls, so the offsets weren't correct.

mvdbeek committed:
Seek to right position, then read

mvdbeek reviewed (approved):
Looks good to me, thanks @annefou, and sorry for missing the right offset in my first suggestion.

mvdbeek committed:
Open grib file in binary mode
